### PR TITLE
[FluidAPI]remove clip api

### PR DESCRIPTION
**Pull request by wj-Mcat**
### PR types

Function optimization

### PR changes

APIs

### Describe

在 clip 模块的迁移工作当中，需要将fluid/clip.py 模块整体迁移到nn/clip.py中，除了三个公共的 api，还有五个内部api ，迁移以下 fluid API

公共 api：
- [x] ClipGradByGlobalNorm
- [x] ClipGradByNorm
- [x] ClipGradByValue

内部 api：
- [x] clip_by_norm
- [x] merge_selected_rows
- [x] get_tensor_from_selected_rows
- [x] ErrorClipByValue
- [x] set_gradient_clip

其中涉及调用的模块比较广泛，所以 file change数量也是比较多。

### CODE REVIEW

File: python/paddle/fluid/tests/unittests/test_clip_op.py [hunk 1]
```diff
@@ -129,15 +129,9 @@ def test_errors(self):
...
-                fluid.layers.clip(x=input_data, min=-1.0, max=1.0)
+                paddle.clip(x=input_data, min=-1.0, max=1.0)
 
             self.assertRaises(TypeError, test_Variable)
-
-            def test_dtype():
-                x2 = fluid.layers.data(name='x2', shape=[1], dtype='int32')
-                fluid.layers.clip(x=x2, min=-1.0, max=1.0)
-
-            self.assertRaises(TypeError, test_dtype)
```
Comment (wj-Mcat) [inline python/paddle/fluid/tests/unittests/test_clip_op.py:139]: 这个单测调整到paddle.clip 之后就过了，所以在最新的 2.0 的 api 里面是支持 int32 的了。 @zoooo0820

File: python/paddle/fluid/tests/unittests/xpu/test_clip_op_xpu.py [hunk 2]
```diff
@@ -131,13 +131,13 @@ def test_errors(self):
...
             def test_Variable():
-                fluid.layers.clip(x=input_data, min=-1.0, max=1.0)
+                paddle.clip(x=input_data, min=-1.0, max=1.0)
 
             self.assertRaises(TypeError, test_Variable)
 
             def test_dtype():
                 x2 = fluid.layers.data(name='x2', shape=[1], dtype='int32')
-                fluid.layers.clip(x=x2, min=-1.0, max=1.0)
+                paddle.clip(x=x2, min=-1.0, max=1.0)
```
Comment (wj-Mcat) [inline python/paddle/fluid/tests/unittests/xpu/test_clip_op_xpu.py:140]: 不知道为什么这里的测试在 xpu 上面会挂掉。

Comment (wj-Mcat) [reply python/paddle/fluid/tests/unittests/xpu/test_clip_op_xpu.py:140]: 这个是非 required 的选项，沟通之后说可以先暂时不管。

File: python/paddle/nn/clip.py [hunk 3]
```diff
@@ -12,9 +12,1088 @@
...
+        attrs={},
+        outputs={"Out": out},
+    )
+    return out
+
+
+@templatedoc()
+def get_tensor_from_selected_rows(x, name=None):
+    """
+    This operator gets tensor data from input with SelectedRows type, and outputs a LoDTensor.
```
Comment (Ligoml) [inline python/paddle/nn/clip.py]:
```suggestion
    Get tensor data from input with SelectedRows type, and outputs a Tensor.
```

Comment (Ligoml) [reply python/paddle/nn/clip.py]: 1. api文档中尽量不要出现op operator等错误表述
2. 飞桨现在已经没有LoDTensor

File: python/paddle/nn/clip.py [hunk 4]
```diff
@@ -12,9 +12,1088 @@
...
+        or x.dtype == core.VarDesc.VarType.BF16
+    ) and _clip_by_global_norm_using_mp_type():
+        return x.astype(core.VarDesc.VarType.FP32)
+    else:
+        return x
+
+
+def _squared_l2_norm(x):
+    r"""
+    This OP returns the squared L2 norm of a tensor.
```
Comment (Ligoml) [inline python/paddle/nn/clip.py]:
```suggestion
    Return the squared L2 norm of a tensor.
```

File: python/paddle/nn/clip.py [hunk 5]
```diff
@@ -12,9 +12,1088 @@
...
+    def __str__(self):
+        raise NotImplementedError()
+
+    def _append_clip_op(self, block, grad_name):
+        raise NotImplementedError()
+
+
+class ErrorClipByValue(BaseErrorClipAttr):
+    r"""
+    Clips tensor values to the range [min, max].
```
Comment (Ligoml) [inline python/paddle/nn/clip.py]:
```suggestion
    Clip tensor values to the range [min, max].
```

File: python/paddle/nn/clip.py (hunk 4 above)
Comment (Ligoml) [inline python/paddle/nn/clip.py]:
```suggestion
    Return the squared L2 norm of a tensor.
```

File: python/paddle/nn/clip.py [hunk 6]
```diff
@@ -12,9 +12,1088 @@
...
+            outputs={'Out': grad},
+        )
+
+        return param, grad
+
+
+@framework.dygraph_not_support
+def set_gradient_clip(clip, param_list=None, program=None):
+    """
+    :api_attr: Static Graph
```
Comment (Ligoml) [inline python/paddle/nn/clip.py]: 这一行删掉，在文档中没什么意义

**Review by Ligoml — COMMENTED** [review]
移出api需要构建一个官网预览，确保api文档在官网正常显示

File: python/paddle/nn/clip.py [hunk 7]
```diff
@@ -12,9 +12,1088 @@
...
+)
+from paddle.tensor.layer_function_generator import templatedoc
+
+__all__ = [
+    'set_gradient_clip',
+    'ErrorClipByValue',
+    'ClipGradByValue',
+    'ClipGradByNorm',
+    'ClipGradByGlobalNorm',
+]
```
Comment (jeff41404) [inline python/paddle/nn/clip.py]: `ClipGradByValue`, `ClipGradByNorm` and `ClipGradByGlobalNorm` are already in \_\_all\_\_ of `paddle/nn/__init__.py`, so not need to be added here. `set_gradient_clip` and `ErrorClipByValue` should not be public API, only for internal use, also not need to be added here.

File: python/paddle/nn/clip.py (hunk 6 above)
Comment (wj-Mcat) [reply python/paddle/nn/clip.py]: done

File: python/paddle/nn/clip.py (hunk 7 above)
Comment (wj-Mcat) [reply python/paddle/nn/clip.py]: done from your suggestion.  you can checkout detail in https://github.com/PaddlePaddle/Paddle/pull/48946/commits/fde5139d7371dc1b550591b857464e3dad1a2fa6

File: python/paddle/fluid/clip.py [hunk 8]
```diff
@@ -30,7 +30,6 @@
 
 __all__ = [
     'set_gradient_clip',
-    'ErrorClipByValue',
     'ClipGradByValue',
     'ClipGradByNorm',
     'ClipGradByGlobalNorm',
```
Comment (jeff41404) [inline python/paddle/fluid/clip.py]: the implementation of these functions and classes are migrated to `paddle/nn/clip.py`, so the list of \_\_all\_\_ should be clean and implementation here should be deleted.

**Comment by jeff41404** [comment]
All cleaning and migration fluid API need to be written in the `Describe` above to facilitate information synchronization.

**Comment by wj-Mcat** [comment]
> All cleaning and migration fluid API need to be written in the `Describe` above to facilitate information synchronization.

Oh, it's my fault.  I will update description later.

File: python/paddle/fluid/clip.py (hunk 8 above)
Comment (wj-Mcat) [reply python/paddle/fluid/clip.py]: This file should & has been removed with your suggestion.

File: python/paddle/nn/__init__.py [hunk 9]
```diff
@@ -19,9 +19,9 @@
 from .layer.container import ParameterList  # noqa: F401
 from ..fluid.dygraph.container import Sequential  # noqa: F401
 
-from .clip import ClipGradByGlobalNorm  # noqa: F401
-from .clip import ClipGradByNorm  # noqa: F401
-from .clip import ClipGradByValue  # noqa: F401
+from paddle.nn.clip import ClipGradByGlobalNorm  # noqa: F401
+from paddle.nn.clip import ClipGradByNorm  # noqa: F401
+from paddle.nn.clip import ClipGradByValue  # noqa: F401
```
Comment (jeff41404) [inline python/paddle/nn/__init__.py]: no need to modify? it is OK to modify this way, but the format is different from others below

File: python/paddle/nn/clip.py [hunk 10]
```diff
@@ -12,9 +12,1078 @@
...
 # limitations under the License.
 
-# TODO: define the functions to clip gradient of parameter
-from ..fluid.clip import ClipGradByGlobalNorm  # noqa: F401
-from ..fluid.clip import ClipGradByNorm  # noqa: F401
-from ..fluid.clip import ClipGradByValue  # noqa: F401
+import copy
+import warnings
 
-__all__ = []
```
Comment (jeff41404) [inline python/paddle/nn/clip.py:20]: no need to modify? we can leave list of \_\_all\_\_ empty

File: python/paddle/nn/__init__.py (hunk 9 above)
Comment (wj-Mcat) [reply python/paddle/nn/__init__.py]: ok，I will revert it to keep it consistent.

File: python/paddle/nn/clip.py (hunk 10 above)
Comment (wj-Mcat) [reply python/paddle/nn/clip.py:20]: ok, I will revert it.

**Review by Ligoml — DISMISSED** [review]
LGTM for docs

**Review by Ligoml — APPROVED** [review]
LGTM for docs

**Review by jeff41404 — APPROVED** [review]
LGTM

**Review by raindrops2sea — APPROVED** [review]

File: python/paddle/nn/clip.py [hunk 11]
```diff
@@ -12,9 +12,1074 @@
...
+        inputs={"X": x},
+        attrs={"max_norm": max_norm},
+        outputs={"Out": out},
+    )
+
+    return out
+
+
+@templatedoc()
+def merge_selected_rows(x, name=None):
```
Comment (XiaoguangHu01) [inline python/paddle/nn/clip.py:88]: 这个函数看起来跟clip没啥关系呢，是不是可以放到其他地方？

File: python/paddle/nn/clip.py [hunk 12]
```diff
@@ -12,9 +12,1074 @@
...
+        type="merge_selected_rows",
+        inputs={"X": x},
+        attrs={},
+        outputs={"Out": out},
+    )
+    return out
+
+
+@templatedoc()
+def get_tensor_from_selected_rows(x, name=None):
```
Comment (XiaoguangHu01) [inline python/paddle/nn/clip.py:127]: 这个函数看起来跟clip关系不大

**Review by XiaoguangHu01 — APPROVED** [review]
可以先合入，然后再看下我留的评论

File: python/paddle/nn/clip.py (hunk 11 above)
Comment (wj-Mcat) [reply python/paddle/nn/clip.py:88]: 此函数在起初迁移工作中的建议位置，我这边也没有作更深入的分析和讨论。 以下函数`get_tensor_from_selected_rows`同理，不过后期可在另外一个 pr 中将其迁移到其他模块当中。

**Review by qili93 — APPROVED** [review]
LGTM for init.py

**Review by phlrain — APPROVED** [review]

**Review by lanxianghit — APPROVED** [review]

**Review by FrostML — APPROVED** [review]

**Pull request merged**